include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,236 @@
+            return *static_cast<Type*>(this);
+        }
+
+        /**
+         * @brief Raises a property change notification event
+         * @param name The name of the property
+         * @return
+         * @details Usage example\n
+         * C++
+         * @code
+         * void MyPage::DoSomething() {
+         *  // modify MyInt
+         *  // MyInt = ...
+         *
+         *  // now send a notification to update the bound UI elements
+         *  RaisePropertyChanged(L"MyInt");
+         * }
+         * @endcode
+        */
+        auto RaisePropertyChanged(std::wstring_view name) {
+            return m_propertyChanged(self(), Xaml_Data_PropertyChangedEventHandler{ name });
+        }
+    protected:
+        winrt::event<Xaml_Data_PropertyChangedEventHandler> m_propertyChanged;
+    };
+
+    /**
+     * @brief Implements a property type with notifications
+     * @tparam T the property type
+     * @details Use the #INIT_NOTIFY_PROPERTY macro to initialize this property in your class constructor. This will set up the right property name, and bind it to the `notify_property_changed_base` implementation.
+    */
+    template<typename T,
+        typename Xaml_Data_PropertyChangedEventHandler = wil::details::Xaml_Data_PropertyChangedEventHandler,
+        typename Xaml_Data_PropertyChangedEventArgs = wil::details::Xaml_Data_PropertyChangedEventArgs>
+    struct single_threaded_notifying_property : single_threaded_rw_property<T> {
+        using Type = T;
+
+        using single_threaded_rw_property<T>::operator();
+
+        void operator()(const T& value) {
+            if (value != this->m_value) {
+                single_threaded_rw_property<T>::operator()(value);
+                if (m_npc) {
+                    (*m_npc)(m_sender, Xaml_Data_PropertyChangedEventArgs{ m_name });
+                }
+            }
+        }
+        template<typename... TArgs>
+        single_threaded_notifying_property(
+            winrt::event<Xaml_Data_PropertyChangedEventHandler>* npc,
+            winrt::Windows::Foundation::IInspectable sender,
+            std::wstring_view name,
+            TArgs&&... args) :
+            single_threaded_rw_property<T>(std::forward<TArgs...>(args)...),
+            m_name(name),
+            m_npc(npc),
+            m_sender(sender)
+        {}
+
+        single_threaded_notifying_property(const single_threaded_notifying_property&) = default;
+        single_threaded_notifying_property(single_threaded_notifying_property&&) = default;
+        std::wstring_view Name() const noexcept { return m_name; }
+    private:
+        std::wstring_view m_name;
+        winrt::event<Xaml_Data_PropertyChangedEventHandler>* m_npc;
+        winrt::Windows::Foundation::IInspectable m_sender;
+    };
+
+    /**
+    * @def INIT_NOTIFY_PROPERTY
+    * @brief use this to initialize a wil::single_threaded_notifying_property in your class constructor.
+    */
+#define INIT_NOTIFY_PROPERTY(NAME, VALUE)  \

```

Comment:
It would be great if there were also type-checked macros for raising events as well as handling them.  Everything is a string so both of those types of code are subject to typo errors that will not be flagged by the compiler.

Reply:
I missed that changing the property automatically raises the event (with the correct name!) so that is covered.  That just leaves event handlers.  This is "nice to have" but certainly non-blocking.

The ask is to have something like this:
```cpp
void OnPropertyChanged(const winrt::IInspectable& /*sender*/, const winrt::PropertyChangedEventArgs& args)
{
    // typo-safe property change handling
    if (args.propertyName() == NAMEOF(MyRuntimeClass, MyPropertyName))
    {
        // Handle property changed
    }

    // unsafe property change handling
    if (args.propertyName() == "MyPropertyNameWithTypo")
    {
        // Will never get called because I can't spell
    }
```

Reply:
I'll send a separate PR for that since this feature is somewhat orthogonal to properties/notifications (though it is useful in the context of). See draft #321